packages/aws-appsync/src/client.js

```diff
@@ -9,14 +9,70 @@
 import 'setimmediate';
 import ApolloClient, { ApolloClientOptions, MutationOptions } from 'apollo-client';
 import { InMemoryCache } from 'apollo-cache-inmemory';
-import { ApolloLink, FetchResult } from 'apollo-link';
+import { ApolloLink, FetchResult, Observable } from 'apollo-link';
 import { HttpLink } from 'apollo-link-http';
 import { getMainDefinition, getOperationDefinition, variablesInOperation } from 'apollo-utilities';
 
 import OfflineCache from './cache/index';
 import { OfflineLink, AuthLink, NonTerminatingHttpLink, SubscriptionHandshakeLink, ComplexObjectLink } from './link';
 import { createStore } from './store';
 
+export const createSubscriptionHanshakeLink = (url, resultsFetcherLink = new HttpLink({ uri: url })) => {
+    return ApolloLink.split(
+        operation => {
+            const { query } = operation;
+            const { kind, operation: graphqlOperation } = getMainDefinition(query);
+            const isSubscription = kind === 'OperationDefinition' && graphqlOperation === 'subscription';
+
+            return isSubscription;
+        },
+        ApolloLink.from([
+            new NonTerminatingHttpLink('subsInfo', { uri: url }, true),
+            new SubscriptionHandshakeLink('subsInfo'),
+        ]),
+        resultsFetcherLink,
+    );
+};
+
+export const createAuthLink = ({ url, region, auth }) => new AuthLink({ url, region, auth });
+
+const passthrough = (op, forward) => (forward ? forward(op) : Observable.of());
+
+export const createAppSyncLink = ({
+    url,
+    region,
+    auth,
+    complexObjectsCredentials,
+    resultsFetcherLink = new HttpLink({ uri: url }),
+}) => {
+    const link = ApolloLink.from([
+        createLinkWithStore((store) => new OfflineLink(store)),
+        new ComplexObjectLink(complexObjectsCredentials),
+        createAuthLink({ url, region, auth }),
+        createSubscriptionHanshakeLink(url, resultsFetcherLink)
+    ].filter(Boolean));
+
+    return link;
+};
+
+export const createLinkWithCache = (createLinkFunc = () => new ApolloLink(passthrough)) => {
+    let theLink;
+
+    return new ApolloLink((op, forward) => {
+        if (!theLink) {
```

Comment:
Why are you checking this ```theLink```?

Reply:
Because I want to create the link only once and re-use it on subsequent link invocations. For this I am creating a [closure](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures)